tests/console/nvme_checks.pm

```diff
@@ -0,0 +1,121 @@
+    assert_script_run("nvme id-ctrl -H /dev/$nvm_test_data->{nvm_char_device} > id_ctrl_output");
+    record_info "Identify Admin Command Output", script_output("cat id_ctrl_output");
+    my $MODE       = '\/dev\/' . $nvm_test_data->{nvm_char_device};
+    my $SN         = $nvm_test_data->{sn};
+    my $MODEL      = $nvm_test_data->{model};
+    my $NAMESPACE  = $nvm_test_data->{namespace_count};
+    my $DISK_SIZE  = $nvm_test_data->{nvm_disk_size};
+    my $BLOCK_SIZE = $nvm_test_data->{nvm_block_size};
+    my $list_regex = qr/${MODE}\s+$SN\s+${MODEL}\s+${NAMESPACE}\s+${DISK_SIZE}\s+GB.*\d+\s+GB\s+${BLOCK_SIZE}/;
+    validate_script_output("nvme list",                                                                     sub { $list_regex });
+    validate_script_output("nvme list-ns /dev/$nvm_test_data->{nvm_char_device} | awk -F ':' '{print $2}'", sub { /$nvm_test_data->{nvm_ns}/ });
+    # the details of this namespace
+    # check LBA Format fields
+    # ms = Metadata Size, 0 means metadata is not supported
+    # lbads = LBA Data Size, in terms of a power of two, so LBA data size is 2⁹ = 512 bytes
+    # rp = Relative Performance, the lower the better
+    validate_script_output("nvme id-ns /dev/$nvm_test_data->{nvm_disk} --namespace-id=$nvm_test_data->{nvm_ns}", sub { m/lbaf\s+0\s+:\s+ms:0\s+lbads:9\s+rp:0\s\(in use\)/ });
+    # show registers
+    my $ret_code = script_run("nvme show-regs -H /dev/$nvm_test_data->{nvm_disk}");
+    record_soft_failure "bsc#1175217 - not permitted shared memory access to get registers" if ($ret_code > 0);
+}
+
+sub _issue_read_command_and_compare_with_dd {
+    my $nvm_test_data = shift;
+    assert_script_run("dd if=/dev/$nvm_test_data->{nvm_disk} of=lba.0.dd bs=512 count=1");
+    # read LBA 0 through nvme utility
+    assert_script_run("nvme read /dev/$nvm_test_data->{nvm_disk} --start-block=0 --block-count=0 --data-size=512 --data=lba.0.read");
+    # Read is opcode 0x02, we are sending command to namespace 0x1, reading 512 bytes
+    # code word 10 and 11 (cdw10 and cdw11) specifies the start of reading in LBA blocks,
+    # and the Bits 15:00 of code word 12 (cdw12) is number of blocks to read (set to 0 to indicate 1 block will be read).
+    assert_script_run("nvme io-passthru /dev/$nvm_test_data->{nvm_char_device} --opcode=0x02 --namespace-id=0x1 --data-len=512 --read --cdw10=0 --cdw11=0 --cdw12=0 -b > lba.0.io");
+    assert_script_run("cmp lba.0.dd lba.0.read", fail_message => "The Admin read command is broken. Results are different between the helper read command and the manual dd");
+    assert_script_run("cmp lba.0.dd lba.0.io", fail_message => "The Submission and Completion queue for Admin commands differs from I/O commands");
+}
+
+sub _check_if_nvme_tools_is_installed {
+    my $ret = zypper_call('se -x -i nvme-cli', exitcode => [0, 104]);
+    if ($ret == 104) {
+        record_soft_failure "bsc#1172866 - nvme-cli is not installed" unless script_run("rpm -qa nvme-cli");
```

Comment:
you don't need the unless anymore here.
First time I see a method which name contains 'if', good practice is to use verbs and names, i.e.: `_check_nvme_tools_installed`.

Reply:
Return code 104 means that the package was not found (EDIT: not in package database) not that the package is not installed. I find it confusing. Also, "rpm -qa nvme-cli" will return 0 whether package is installed or not.

Reply:
then, if return code is 104, package cannot be installed so next line will fail.

Reply:
I think this is not correct. `104 - ZYPPER_EXIT_INF_CAP_NOT_FOUND
           Returned by the install and the remove command in case any of the arguments does not match any of the
           available (or installed) package names or other capabilities.`

indeed unless has to be removed. i dont know why i left it :)

Reply:
I think it was correct, as we are searching with `-i` flag, so we are searching only on the installed packages and it is not found.

Reply:
Yes the use of return code was correct. But it does not ensure that the package is installed or not. This is the best way I come up as i tried also other things. For example a combination of zypper and rpm. But it looks like uncessary complicated. So I am not intended to make any change on that.

Reply:
I missed the -i flag, now makes sense. I think you should apply Joaquin's suggestion to use rpm -qa | grep nvme-cli here, it's simple and should work I guess.

Reply:
I think it is fine as it is.

Reply:
Please, revert to solution where ret code are checked and library is used (or even with rpm in the suggested way). Right now you are checking with regex for 'i' which is not necessary due to the output will always contain 'i' as you filter with zypper by installed packages.
It is wrong to write a condition in the code that always is true.

Reply:
For me it still looks wrong, so now you search for installed package and if the package is installed you get the 'i' and all fine, but if it is not installed it is not listed and you grab in the output 'whatever' (for example for me grab it 'No' from 'No matching items found). It is unclear from the code what this does comparing with the previous solution. It works but it is weaker that the other two approaches suggested.

Reply:
+1 ^^, the return code actually made sense with -i, but the unless had no utility.

Reply:
Yea i know. Simple solution. If `+i` means is installed, print info and continue.

Reply:
Sorry i missed this

Reply:
@rwx788 wdyt?

Reply:
reverted as it was in first place. too much confusion. resolved